sonarqube/datadog_checks/sonarqube/discovery_filter.py

```diff
@@ -0,0 +1,135 @@
+# They may be of type string or key/value.
+# If its key is found in the list, the tuple formed by the key/value or
+# key/None in the case of being a string type will be returned.
+# 'discovery': It is the part of the configuration that will allow the 'autodiscover' feature within a list of values.
+# It works by comparing patterns following the following criteria:
+# The elements of the list to be filtered that match any regular expression included in the 'include' key of 'discovery'
+# (or default_include list) and do not match any of the regular expressions of the 'exclude' key of 'discovery'
+# (or default_exclude list).
+# The include patterns will be processed in order
+# The maximum number of elements to be returned by the 'autodiscovery' part will be the one indicated in the 'limit' key
+# (or default_limit)
+# Different examples of these configurations can be seen in test_unit_projects.py inside tests
+class DiscoveryFilter:
+    def __init__(
+        self,
+        name,
+        log,
+        config,
+        mandatory=True,
+        items_id='keys',
+        default_limit=10,
+        default_include=None,
+        default_exclude=None,
+    ):
+        self._log = log
+        self._items_id = items_id
+        self._config = config
+        self._default_limit = default_limit
+        self._default_include = [] if default_include is None else default_include
+        self._default_exclude = [] if default_exclude is None else default_exclude
+        self._log.debug('\'%s\' config: %s', name, self._config)
+        if self._config is None and mandatory:
+            raise ConfigurationError('\'{}\' setting must be defined'.format(name))
+
+    def match(self, items):
+        self._log.debug('trying to match: %s', items)
+        self._log.debug('self._config: %s', self._config)
+        matched_items = self._match_items(items)
+        matched_items.extend(self._match_discovery(matched_items, items))
+        return matched_items
+
+    def _match_items(self, items):
+        matched_id = []
+        self._log.debug('matching items')
+        self._log.debug('items: %s', items)
+        if self._config is not None and isinstance(self._config, dict):
+            config_items_id = self._config.get(self._items_id, None)
+            self._log.debug('config_items_id: %s', config_items_id)
+            if config_items_id:
+                for id_item in config_items_id:
+                    id_key, id_value = None, None
+                    if isinstance(id_item, dict):
+                        id_key = list(id_item)[0]  # iterating over a dictionary implicitly iterates only over its keys
+                        id_value = list(id_item.values())[0]
+                    elif isinstance(id_item, str):
+                        id_key = id_item
+                    if id_key not in [key for key, _ in matched_id]:
+                        if id_key in items:
+                            self._log.debug('\'%s\' item matched', id_key)
```

Comment:
For quoting use the backtick like in Markdown ``` ` ```

Reply:
Done